### ISSUE TITLE
Unnecessary assignment incorrectly marks the variable as unnecessary

### ISSUE THREAD

**Issue by svick** [dotnet/roslyn#47784]
**Version Used**: VS 2019 16.8.0 Preview 2.1

**Steps to Reproduce**:

Have code like the following open in VS:

```c#
int F(bool condition)
{
    int variable = 0;

    if (condition)
        variable = 1;
    else
        variable = 2;

    return variable;
}
```

**Expected Behavior**: IDE0059 "Unnecessary assignment of a value" highlights the assignment of 0 to `variable` as unnecessary:

```c#
int variable[| = 0|];
```

**Actual Behavior**:

The variable itself is highlighted as unnecessary by IDE0059, indicating that the variable should be removed, not just the assignment:

```c#
int [|variable|] = 0;
```

![](https://user-images.githubusercontent.com/287848/93471181-511fc780-f8f3-11ea-83c4-3370f294cb69.png)

If you actually run the code fix, it does the right thing, i.e. it only removes the assignment (and also the empty line, for some reason):

![](https://user-images.githubusercontent.com/287848/93471568-d0150000-f8f3-11ea-8ba6-c5bfc7a97e7d.png)

**Comment by Youssef1313** [comment]
To make sure I understand this correctly, the problem is only with what's being faded-out right? i.e. `= 0` should be faded instead of `variable`.

**Comment by svick** [comment]
@Youssef1313 Yeah, exactly. ("Faded out" is probably a better term than "marked as unnecessary", but that didn't come to my mind when I was writing this.)

**Comment by Youssef1313** [comment]
https://github.com/dotnet/roslyn/blob/e16d3906a6d6a560da72144a942dbac7c8dc2e11/src/Analyzers/CSharp/Analyzers/RemoveUnusedParametersAndValues/CSharpRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs#L62

Probably this will be a good start in debugging.
I'll give it a try soon.

**Comment by Youssef1313** [comment]
The code fix unfortunately relies on the primary diagnostic location. So I think fixing the diagnostic location will require tweaks to the codefix too.

@mavasani @CyrusNajmabadi Is there a way to explicitly set the fading span but keep the primary diagnostic location the same to avoid doing many changes?

**Comment by mavasani** [comment]
I believe this is by design. Note that by attempting to fade some other location, we are trading one problem for another.

1. If we decide to fade the entire `= expr`, the expression on right might even span across multiple lines, so fading the entire equals value clause is also not really desirable. Additionally, user might feel the evaluation of `expr` is unnecessary, which is not always true as evaluating `expr` might have side effects. The code fix will not suggesting removing the assignment for such a case, but replace the left side with a discard.
2. If we decide to fade only the `=`, it will hardly be visible, if at all. Additionally, I am not sure how do we handle the unnecessary assignment for an out variable like below:
![image](https://user-images.githubusercontent.com/10605811/93493026-50ba0900-f8c0-11ea-9abd-168bb7ab506d.png)

The current approach of fading the left hand side is designed to indicate an `unnecessary write`, not an `unnecessary variable`. I don't believe fading any other location would help here.

**Comment by svick** [comment]
@mavasani 

> The code fix will not suggesting removing the assignment for such a case, but replace the left side with a discard.

I think that the faded part should be whatever is going to be removed by the code fix, which should solve this issue.

> If we decide to fade only the `=`, it will hardly be visible, if at all.

I agree, fading just the `=` is not better.

> The current approach of fading the left hand side is designed to indicate an "unnecessary write" or "unnecessary store", not an "unnecessary variable".

I think that works for an assignment, but not for a local variable declaration. There, fading the variable name does not indicate "unnecessary write" to me at all.

**Comment by sharwell** [comment]
Duplicate of #36202
